File: Models/Stics_soil_temperature/Tempamp.f90
MODULE temp_amp_mod
   IMPLICIT NONE
CONTAINS
   !%%CyML Model Begin%%
   SUBROUTINE model_temp_amp(min_temp, &
                            max_temp, &
                            temp_amp)
      IMPLICIT NONE

      REAL, INTENT(IN) :: min_temp
      REAL, INTENT(IN) :: max_temp
      REAL, INTENT(OUT) :: temp_amp

      !%%CyML Description Begin%%
      !- Name: temp_amp -Version: 1.0, -Time step: 1
      !- Description:
      !            * Title: temp_amp model
      !            * Author: STICS
      !            * Reference: doi:http://dx.doi.org/10.1016/j.agrformet.2014.05.002
      !            * Institution: INRAE
      !            * Abstract: Calculates temperature amplitude
      !- inputs:
      !            * name: min_temp
      !                          ** description : current minimum temperature
      !                          ** inputtype : variable
      !                          ** variablecategory : exogenous
      !                          ** datatype : DOUBLE
      !                          ** default : 0.0
      !                          ** min : -50.0
      !                          ** max : 50.0
      !                          ** unit : degC
      !                          ** uri :
      !            * name: max_temp
      !                          ** description : current maximum temperature
      !                          ** inputtype : variable
      !                          ** variablecategory : exogenous
      !                          ** datatype : DOUBLE
      !                          ** default : 0.0
      !                          ** min : -50.0
      !                          ** max : 50.0
      !                          ** unit : degC
      !                          ** uri :
      !- outputs:
      !            * name: temp_amp
      !                          ** description : current temperature amplitude
      !                          ** variablecategory : state
      !                          ** datatype : DOUBLE
      !                          ** min : 0.0
      !                          ** max : 100.0
      !                          ** unit : degC
      !                          ** uri :
      !%%CyML Description End%%

      !%%CyML Compute Begin%%
      temp_amp = max_temp - min_temp
      !%%CyML Compute End%%

   END SUBROUTINE model_temp_amp
   !%%CyML Model End%%
END MODULE temp_amp_mod

File: Models/Stics_soil_temperature/src/cpp/Stics_soil_temperature/Temp_amp.cpp
#define _USE_MATH_DEFINES
#include <cmath>
#include <iostream>
#include <vector>
#include <string>
#include <numeric>
#include <algorithm>
#include <array>
#include <map>
#include <tuple>
#include "temp_amp.h"
using namespace Stics_soil_temperature;
temp_amp::temp_amp() {}
void temp_amp::Calculate_Model(soil_tempState &s, soil_tempState &s1, soil_tempRate &r, soil_tempAuxiliary &a, soil_tempExogenous &ex)
{
    //- Name: temp_amp -Version: 1.0, -Time step: 1
    //- Description:
    //            * Title: temp_amp model
    //            * Authors: None
    //            * Reference: doi:http://dx.doi.org/10.1016/j.agrformet.2014.05.002
    //            * Institution: INRAE
    //            * ExtendedDescription: None
    //            * ShortDescription: None
    //- inputs:
    //            * name: min_temp
    //                          ** description : current minimum temperature
    //                          ** inputtype : variable
    //                          ** variablecategory : exogenous
    //                          ** datatype : DOUBLE
    //                          ** max : 50.0
    //                          ** min : -50.0
    //                          ** default : 0.0
    //                          ** unit : degC
    //            * name: max_temp
    //                          ** description : current maximum temperature
    //                          ** inputtype : variable
    //                          ** variablecategory : exogenous
    //                          ** datatype : DOUBLE
    //                          ** max : 50.0
    //                          ** min : -50.0
    //                          ** default : 0.0
    //                          ** unit : degC
    //- outputs:
    //            * name: temp_amp
    //                          ** description : current temperature amplitude
    //                          ** datatype : DOUBLE
    //                          ** variablecategory : state
    //                          ** max : 100.0
    //                          ** min : 0.0
    //                          ** unit : degC
    double min_temp = ex.getmin_temp();
    double max_temp = ex.getmax_temp();
    double temp_amp;
    temp_amp = max_temp - min_temp;
    s.settemp_amp(temp_amp);
}